Issue 3057988 Automatically fix deprecated layout text

jonathan1055 opened the pull request:
https://www.drupal.org/project/coder/issues/3057988

jonathan1055 commented:
The test failure is one PHPstan problem that exists in the 8.x-3.x branch anyway. The code is clean using phpcs and the modified tests pass

klausi commented on coder_sniffer/Drupal/Sniffs/Commenting/DeprecatedSniff.php:
```diff
@@ -75,20 +74,72 @@ public function process(File $phpcsFile, $stackPtr)
+                    // This problem is a drupal core file and is fixable.
+                    if (empty($matchesFix[1]) === false) {
+                        // Verify that its acceptable to remove the text in [1].
+                        echo('>>> File: '.$phpcsFile->path.' line '.$tokens[($stackPtr)]['line'].PHP_EOL);
```
hm, I think we should not use echo() in Coder. All output should be done with addError() or addWarning().

Or this for debugging? Then we should hide this behind $this->debug check. See ScopeIndentSniff for example.

jonathan1055 replied:
It was for debug, and useful to keep in, so I will use your hint about $this->debug. Is it ok to call the runtime parameter `deprecated_debug` following the similarly named `scope_indent_debug` ?

klausi commented on coder_sniffer/Drupal/Sniffs/Commenting/DeprecatedSniff.php:
```diff
@@ -75,20 +74,72 @@ public function process(File $phpcsFile, $stackPtr)
+                    }
+
+                    $fix = $phpcsFile->addFixableError($error, $key, 'IncorrectTextLayoutFixable', [$fullText]);
+                    $phpcsFile->fixer->replaceToken($key, $correctedText);
```
only if $fix is true we are allowed to replace stuff. See other sniffs how they fix things.

jonathan1055 replied:
OK, will do

klausi commented on coder_sniffer/Drupal/Sniffs/Commenting/DeprecatedSniff.php:
```diff
@@ -75,20 +74,72 @@ public function process(File $phpcsFile, $stackPtr)
+                        echo('>>> WARNING: length of $correctedText = '.strlen($correctedText).' (so line will exceed 80)'.PHP_EOL);
+                    }
+
+                    $fix = $phpcsFile->addFixableError($error, $key, 'IncorrectTextLayoutFixable', [$fullText]);
```
we should not have different error codes for fixable errors. Just use 1 code.

jonathan1055 replied:
OK

klausi commented on coder_sniffer/Drupal/Sniffs/Commenting/DeprecatedSniff.php:
```diff
@@ -75,20 +74,72 @@ public function process(File $phpcsFile, $stackPtr)
+                // Get just the first line of the text.
+                $key   = array_keys($textItems)[0];
+                $text1 = $textItems[$key];
+                preg_match('/^(.*)(as of|in) (drupal|)( |:|)+([\d\.\-xdev\?]+)(,| |. |)(.*)(removed|removal)([ |from|before|in|the]*) (drupal|)( |:|)([\d\-\.xdev]+)( |,|$)+(?:release|)(?:[\.,])*(.*)$/i', $text1, $matchesFix);
```
we should not hard-code "drupal" here, right?

jonathan1055 replied:
Thanks for all the review comments. Very helpful. I will start with this one, as it affects how I proceed with the others. If we do not hard-code 'drupal' here then it would mean catering for contrib within the fixable section. I guess that preg_match could be altered to allow any word+undercore instead of 'drupal' but that would mean later testing to see if it was 'drupal' or blank (in which case the fix is for core) or some other word (in which case the fix is for contrib). The version checking would have to be different for contrib, but that could be done. 

If we think this is worth it, I will try to make the changes. But I'm not sure how much deprecation there is in contrib, so I was happy initially to exclude it from automatic fixing. 

The hard-coding of 'drupal' at that line effectively says "I am only attempting to provide auto-fixes for core problems". If I made that clearer in the comments would that be an acceptable solution? I think the alternative (if you do not want 'drupal' hard-coded there) would be for  me to re-write the whole patch to remove the auto-fixable part completely. But it does need the new sniff to cater for missing 'extra info' so there would be some work to do.

klausi replied:
Yep, hard coding "drupal" just for the fixable error as first step would be acceptable I think. Please add code comments to explain that.

Thank you!

jonathan1055 commented:
I have added a new DeprecatedUnitTest.inc.fixed test to verify the auto-fixes, improved the comments, added a debug parameter and fixed the actual code for coding standards. My travis test runs green, with no phpstan or phpcs warnings either.

Unfortunately I had merged in from 8.x-3.x the other recent commits, so they are also showing above. My new changes are all squashed down into one commit, the last one shown above. Is that acceptable for you to work with?

jonathan1055 commented on coder_sniffer/Drupal/Sniffs/Commenting/DeprecatedSniff.php:
```diff
@@ -198,19 +198,24 @@ public function process(File $phpcsFile, $stackPtr)
+        // this is covered by FunctionComment.EmptySees sniff. It also causes
+        // an odd message with '<?php' shown for the @see url content.
+        if (!empty($string)) {
+            // Check the format of the @see url.
```
This is just a simple change, wrapping the url check in a conditional, to avoid doing anything when there is nothing to check. This is covered by FunctionComment.EmptySees anyway.

jonathan1055 replied:
Reverted this change, as the other sniff is only for functions, so was not always being shown. Instead cater for blank @see url by setting $crLink to blank. Two commits squashed together for ease of review.

klausi commented on README.md:
```diff
@@ -129,7 +129,7 @@ Editors:
 
 
-## Automated Testing (PHPUnit + PHPCS)
+## Automated Testing (PHPUnit + PHPCS + PHPStan)
```
Unrelated change - looks like your merged  in the 8.x-3.x branch wrong?

jonathan1055 replied:
I have now pushed a nice clean branch. Only the four files changed. I have left it as two commits, the first is the same a patch 16 which you reviewed. So you can see the changes I made following review.

jonathan1055 commented on coder_sniffer/Drupal/Test/Commenting/DeprecatedUnitTest.inc:
```diff
@@ -2,9 +2,9 @@
 
 /**
  * @file
+ * This file contains test cases for DeprecatedSniff.
```
These changes are to make the test file clean for all coding standards (but I think on reflection this was unnecessary as the test directories are ignored. Oh well ...)

klausi commented on coder_sniffer/Drupal/Sniffs/Commenting/DeprecatedSniff.php:
```diff
@@ -114,15 +200,22 @@ public function process(File $phpcsFile, $stackPtr)
+            // If matches[4] is not blank it means that the url is correct but
+            // it ends with a period. This is covered by the generic fixable
+            // sniff FunctionCommentSniff.SeePunctuation so allow that sniff to
+            // report it (and fix it).
```
This is an empty if() statement now and should be removed?

jonathan1055 commented:
Thanks for commiting. I can make a new PR for removing the empty `if` statement.

If I had known that the new tag 8.x-3.8 was imminent I would have worked on this a few days earlier to get it in, so that Core can start using the sniffs. Do you have a plan for when 3.9 will be released?

klausi commented:
Usually I make releases every 3 months. but if you need to get a release out earlier for drupal core I think we can do it anytime. Best you do a little bit of testing in the next days / weeks and then we could do a new release.

jonathan1055 commented:
Thanks for the info. I may try to work on the trigger_error auto correction too, and then they can both get into the next release. Those two sniffs are closely related.
